src/dscanner/analysis/useless_initializer.d

```diff
@@ -26,301 +19,242 @@ Limitations:
+	override void visit(AST.StructDeclaration structDecl)
 	{
-		if (_inTest)
+		if (inTest || structDecl.ident is null)
 			return;
 
-		assert(_inStruct.length > 1);
+		string structName = cast(string) structDecl.ident.toString();
+		if (isNestedStruct())
+			structName = structStack[$ - 1] ~ "." ~ structName;
 
-		const string structName = _inStruct[$-2] ?
-			_structStack.back() ~ "." ~ decl.name.text :
-		  decl.name.text;
+		bool isDisabled = (structDecl.storage_class & STC.disable) != 0;
+		visitedStructs[structName] = StructInfo(structName, !isDisabled, true);
+		structStack ~= structName;
+		super.visit(structDecl);
 
-		_structStack.insert(structName);
-		_structCanBeInit[structName] = false;
-		_atDisabled.insert(false);
-		decl.accept(this);
-		_structStack.removeBack();
-		_atDisabled.removeBack();
+		visitedStructs[structName].isBeingVisited = false;
+		structStack.length--;
 	}
 
-	override void visit(const(Declaration) decl)
+	private bool isNestedStruct()
 	{
-		_inStruct.insert(decl.structDeclaration !is null);
-
-		with (noLint.push(NoLintFactory.fromDeclaration(decl)))
-			decl.accept(this);
+		if (structStack.length >= 1)
+			return visitedStructs[structStack[$ - 1]].isBeingVisited;
 
-		if (_inStruct.length > 1 && _inStruct[$-2] && decl.constructor &&
-			((decl.constructor.parameters && decl.constructor.parameters.parameters.length == 0) ||
-			!decl.constructor.parameters))
-		{
-			_atDisabled[$-1] = decl.attributes
-				.canFind!(a => a.atAttribute !is null && a.atAttribute.identifier.text == "disable");
-		}
-		_inStruct.removeBack();
+		return false;
 	}
 
-	override void visit(const(Constructor) decl)
+	override void visit(AST.CtorDeclaration ctorDeclaration)
 	{
-		if (_inStruct.length > 1 && _inStruct[$-2] &&
-			((decl.parameters && decl.parameters.parameters.length == 0) || !decl.parameters))
-		{
-			const bool canBeInit = !_atDisabled[$-1];
-			_structCanBeInit[_structStack.back()] = canBeInit;
-			if (!canBeInit)
-				_structCanBeInit[_structStack.back()] = !decl.memberFunctionAttributes
-					.canFind!(a => a.atAttribute !is null && a.atAttribute.identifier.text == "disable");
-		}
-		decl.accept(this);
-	}
+		super.visit(ctorDeclaration);
 
-	// issue 473, prevent to visit delegates that contain duck type checkers.
-	override void visit(const(TypeofExpression)) {}
+		bool isDefaultCtor = ctorDeclaration.getParameterList().length() == 0;
 
-	// issue 473, prevent to check expressions in __traits(compiles, ...)
-	override void visit(const(TraitsExpression) e)
+		if (structStack.length == 0 || !isDefaultCtor)
+			return;
+
+		auto structName = structStack[$ - 1];
+		if (!visitedStructs[structName].isBeingVisited || !visitedStructs[structName].shouldErrorOnInit)
+			return;
+
+		bool isDisabled = (ctorDeclaration.storage_class & STC.disable) != 0;
+		visitedStructs[structName].shouldErrorOnInit = !isDisabled;
+	}
+
+	override void visit(AST.VarDeclaration varDeclaration)
```

Comment:
Please try to find smaller names that are congruent with the dmd codebase. For example, `vd` is consistently used for variable declarations variables throughout dmd. If that seems not expressive enough, you can use varDecl. Having smaller if conditions in length is eye candy.

Reply:
I can shorten it, however for a function this long a longer name is better - I prefer shorter names in local / small scopes. `vd` is an awful name though